### PR TITLE
APIv4 - Support "relative date ranges" with the "IS" operator

### PR DESCRIPTION
**Pull request by totten**
Before
------

In APIv4, there's no good way to use the "relative date ranges" (e.g. `previous.month`, `previous.year`, `this.fiscal_year`, `next.week`).

After
-----

APIv4 supports two new query operators, `IS` and `IS NOT`. The `IS` operator is loosely similar to the SQL `IS` operator in that it accepts symbolic quasi-values (eg `myfield IS null`). This flavor of `IS` accepts `null` as well as the relative date ranges, e.g.

```php
civicrm_api4("Activity", "get", [
  "where" => [["activity_date_time", "IS", "previous.year"]],
  "select" => ["id","activity_date_time"],
  "limit" => 10,
]);
```

### CODE REVIEW

**Comment by totten** [comment]
Aside: I originally drafted this [using an API subscriber to translate the `$where` clauses](https://github.com/civicrm/civicrm-core/commit/a62d8b202a1a2ed5b88eefaab5c30438d090cfa6). That worked OK on simple queries, and it was nice to be a single file, but it would've taken more boilerplate to correctly handle `OR`/`NOT` expressions. The current updates the query-builder, and it seems overall a bit more maintainable.

**Comment by eileenmcnaughton** [comment]
@colemanw are you OK with this - it needs a test. I'm inclined to get it in before the rc is cut if there is a test but I imagine @totten wants feedback first

File: CRM/Core/DAO.php [hunk 1]
```diff
@@ -2734,6 +2734,27 @@ public static function createSQLFilter($fieldName, $filter, $type = NULL, $alias
             }
             return $escapedCriteria;
 
+          // Special purpose "IS" operator
```
Comment (eileenmcnaughton) [inline CRM/Core/DAO.php:2737]: Can we expand this comment a little?

Comment (totten) [reply CRM/Core/DAO.php:2737]: Yup, rewrote the comment 👍

**Comment by totten** [comment]
@colemanw @eileenmcnaughton - I've made more updates:

* Support both DAO-based and array-based entities.
* Support negation (e.g. `['created_date', 'IS NOT', 'this.year']`)
* Include test coverage with a few permutations of DAO/array-entities and affirmative/negative queries.

So I've removed the WIP flag - should be more reviewable now.

**Comment by totten** [comment]
jenkins, test this please

File: CRM/Core/DAO.php [hunk 2]
```diff
@@ -2769,6 +2778,8 @@ public static function acceptedSQLOperators() {
       'NOT IN',
       'BETWEEN',
       'NOT BETWEEN',
+      'IS',
+      'IS NOT',
```
Comment (colemanw) [inline CRM/Core/DAO.php:2782]: I feel reluctant about adding this to CRM_Core_DAO.
This will affect APIv3 not just v4, and probably some other things as well like search builder.
Probably better to create our own list of accepted operators for APIv4.

File: CRM/Core/DAO.php [hunk 3]
```diff
@@ -2734,6 +2734,15 @@ public static function createSQLFilter($fieldName, $filter, $type = NULL, $alias
...
             return $escapedCriteria;
 
+          // The "IS" operator is like an alias; it can be rewritten into other criteria.
+          case 'IS':
+          case 'IS NOT':
+            $newFilter = \Civi\Api4\Utils\CoreUtil::rewriteIsCriteria($fieldName, $operator, $criteria);
+            if ($newFilter !== NULL) {
+              return self::createSQLFilter($fieldName, $newFilter, $type, $alias, $returnSanitisedArray);
+            }
+            return NULL;
```
Comment (colemanw) [inline CRM/Core/DAO.php:2744]: Again I'm not sure that adding this to CRM_Core_DAO is a good idea.
It might be better for APIv4 to depart from using the `CRM_Core_DAO::createSQLFilter` function or at least wrap it.
There are several ways in which APIv4 is diverging from the assumptions in this function, and that will be even more so if we implement stuff like string expressions in the WHERE clause, which I do think we should implement soon.

File: Civi/Api4/Utils/CoreUtil.php [hunk 4]
```diff
@@ -73,4 +73,34 @@ public static function getApiNameFromTableName($tableName) {
...
+   *   Ex: 'activity_date_time'
+   * @param string $operator
+   *   Ex: 'IS' or 'IS NOT'
+   * @param string $criteria
+   *   Ex: ['IS' => 'previous.month']
+   * @return array|NULL
+   *   Array(string $newOperator, mixed $newCriteria).
+   *   Ex: ['BETWEEN' => ['2020-05-01', '2020-06-01']]
+   */
+  public static function rewriteIsCriteria($fieldName, $operator, $criteria) {
```
Comment (colemanw) [inline Civi/Api4/Utils/CoreUtil.php:90]: Would this function be better situated in `\Civi\Api4\Utils\FormattingUtil`?

File: Civi/Api4/Utils/CoreUtil.php [hunk 5]
```diff
@@ -73,4 +73,34 @@ public static function getApiNameFromTableName($tableName) {
...
+    if ($criteria === 'null' || $criteria === 'NULL') {
+      return $affirm ? ['IS NULL' => ''] : ['IS NOT NULL' => ''];
+    }
+
+    $relDateFilters = \CRM_Core_OptionGroup::values('relative_date_filters');
+    if (isset($relDateFilters[$criteria])) {
+      list ($dateFrom, $dateTo) = \CRM_Utils_Date::getFromTo($criteria, NULL, NULL);
+      $newOp = $affirm ? 'BETWEEN' : 'NOT BETWEEN';
+      return [$newOp => [\CRM_Utils_Date::mysqlToIso($dateFrom), \CRM_Utils_Date::mysqlToIso($dateTo)]];
+    }
```
Comment (colemanw) [inline Civi/Api4/Utils/CoreUtil.php:101]: I don't think it's enough to compare the string input to the date filter option list and guess whether or not we're trying to use a relative date filter. What if an activity has the subject "this.year" and we want to search for that activity by subject?

I think this function needs at minimum the entire array of field metadata not just the field name, so it can first check the data type before assuming it's a date.

File: Civi/Api4/Utils/CoreUtil.php [hunk 6]
```diff
@@ -73,4 +73,34 @@ public static function getApiNameFromTableName($tableName) {
...
+
+    return NULL;
```
Comment (colemanw) [inline Civi/Api4/Utils/CoreUtil.php:103]: Instead of returning NULL wouldn't we default to replacing IS with = and IS NOT with !=?

**Review by colemanw — CHANGES_REQUESTED** [review]

**Comment by colemanw** [comment]
Cross linking to https://lab.civicrm.org/dev/event/-/issues/37 so we can keep that in mind. I think the IS operator can be useful for more than just relative dates. We could also consider adding another operator called "CONTAINS" for searching serialized fields.

File: Civi/Api4/Utils/CoreUtil.php (hunk 4 above)
Comment (totten) [reply Civi/Api4/Utils/CoreUtil.php:90]: Yeah, I wasn't really sure. `FormattingUtil` seems just as well.

File: Civi/Api4/Utils/CoreUtil.php (hunk 5 above)
Comment (totten) [reply Civi/Api4/Utils/CoreUtil.php:101]: > What if an activity has the subject "this.year" and we want to search for that activity by subject?

We came into this issue with a definition of the behavior for `['some_date', 'IS', 'this.year']`. However, we didn't explicitly discuss the behavior of `IS` in other scenarios. It could be:

1. Follow in the mold of SQL `IS` and perform basic equality matching on a user-string (same as `=`)
2. Unsupported - throw an error if one tries to use it differently.

There is definitely a learnability benefit of following the SQL `IS` semantics. However, as I think you're saying, things get ambiguous in this case... precisely because it's hard to distinguish a logical symbol (e.g. `"this.year"`) from a literal value (e.g. `"Hello world"`). SQL (as a DSL) can make affordances whenever they want to distinguish logical symbols and literal values (`null` vs `"null"`), but APIv4 operates within the tighter constraints of JSON/PHP-array ...  we can't introduce logical-symbols.

The approach in the current patch aims to resolve ambiguity by making a strict split: `=` matches literal-values, and `IS` matches on a logical symbol. So:

* `['subject', '=', 'this.year']` ==> Match the `subject` against the literal string `this.year`
* `['activity_date_time', 'IS', 'this.year']` ==> Match the `activity_date_time` using the logical symbol `this.year`
* Never the twain shall meet

My gut says that's overall simplest... i.e. no ambiguity, easy to implement.

But it breaks the intuition for people who know SQL... since SQL `IS` can take literal-value operands. We could either:

* Choose a different operator name (`=@` or `=$` or `$=` or somesuch?) to avoid confusion with SQL `IS`
* Generalize APIv4's approach to logical-symbols or pragmas.

    In this case, maybe we look at the other places where API currently uses logical-symbols or processing pragmas:

    * For chaining, the nested API call can use `$value.foo` to reference the prior result. (As it happens, this notation is ambiguous... because `'subject'=>'$value.foo'` could be a logical-symbol or literal-value)
    * For user/contact mapping, there's `@user:alice`. (As it happens, this notation is not problematic... because it's used in numerical contexts... and alphanumerical string necessarily requires evaluation/conversion.)

    Both of those suffer the problem of *obscurity*. The obscurity is amplified by the fact that you cannot write general docs for the API's expression-language -- each notation is bespoke to the field/value/use-case

    If we wanted, we could make the "relative date ranges" follow one of the previous examples closer. Ex: Bless `@` or`$` as the APIv4 sigil:

    | With `@` sigil in the value | With `$` sigil in the value |
    | -- | -- |
    | `['description' => '@value:display_name']` | `['description' => '$value.display_name']` |
    | `['id','=','@user:alice']` | `['id', '=', '$user.alice']` |
    | `['activity_date_time','=','@date:this.year']` | `['activity_date_time','=','$date.this.year']` |

    Any time you see an `@` or `$`, that would mean "the API preprocessor will do magic stuff on this thing." This is nice because it's consistent... and it can be adapted to many different uses within an API-call (get/create/update/delete; select/where/groupBy). But of course, then you have ambiguity for string-values that use `@` or `$`!

    We might resolve the ambiguity by moving the sigil to a different part of the array, eg

    | With `@` postfix in the key |
    | -- |
    | `['@description' => 'value.display_name]` |
    | `['@id', '=', 'user.alice']` |
    | `['@activity_date_time', '=', 'this.year']` |

    It feels a little funny, but I'm pretty sure resolves the ambiguity and works in many different use-cases.

So again... I think this PR (*using different operators for literal-values vs logical-symbols*) probably the simpler play. However, if the others seem higher value, then can we continue digging in...

File: Civi/Api4/Utils/CoreUtil.php (hunk 6 above)
Comment (totten) [reply Civi/Api4/Utils/CoreUtil.php:103]: In keeping with the comment above... the idea in the current PR is to avoid ambiguity by strictly spliting `=` (match by value) and `IS` (match by logical symbol).  Returning `NULL` here means that it will throw an exception if the logical-symbol is unrecognized.

If `IS` were to fallback to `=`, then `IS` would have ambiguity on values like `'this.year'`.

**Comment by colemanw** [comment]
I had another thought about this...
In the scope of this PR, I don't think we actually need a special operator, because date fields are one of the few types of fields that actually preprocess the input, treating it as a date expression rather than a string. So the api can already do things like `activity_date_time = 'now - 1 week'` and it will convert that using `strtotime()`. We could just expand that handling to also deal with range expressions.

**Comment by eileenmcnaughton** [comment]
@colemanw @totten what is the status on this?

**Comment by eileenmcnaughton** [comment]
@colemanw @totten is this proceeding?
